src/spatialdata/_core/query/relational_query.py

```diff
@@ -163,42 +163,60 @@ def _filter_table_by_elements(
     -------
     The filtered table (eventually with reordered rows), or None if the input table was None.
     """
-    assert set(elements_dict.keys()).issubset({"images", "labels", "shapes", "points"})
-    assert len(elements_dict) > 0, "elements_dict must not be empty"
-    assert any(
-        len(elements) > 0 for elements in elements_dict.values()
-    ), "elements_dict must contain at least one dict which contains at least one element"
-    if table is None:
+
+    def _validate_elements_dict(elements_dict: dict[str, dict[str, Any]]) -> None:
+        assert set(elements_dict.keys()).issubset({"images", "labels", "shapes", "points"})
+        assert len(elements_dict) > 0, "elements_dict must not be empty"
+        assert any(
+            len(elements) > 0 for elements in elements_dict.values()
+        ), "elements_dict must contain at least one dict which contains at least one element"
+
+    def _get_table_keys(table: AnnData) -> tuple[str, str]:
+        return (
+            table.uns[TableModel.ATTRS_KEY][TableModel.REGION_KEY_KEY],
+            table.uns[TableModel.ATTRS_KEY][TableModel.INSTANCE_KEY],
+        )
+
+    def _get_element_instances(element: SpatialElement) -> ArrayLike | None:
+        if get_model(element) in [Labels2DModel, Labels3DModel]:
+            if isinstance(element, DataArray):
+                instances = da.unique(element.data).compute()
+            else:
+                assert isinstance(element, DataTree)
+                v = element["scale0"].values()
+                assert len(v) == 1
+                xdata = next(iter(v))
+                instances = da.unique(xdata.data).compute()
+            return np.sort(instances)
+        if get_model(element) == ShapesModel:
+            return element.index.to_numpy()
+        if get_model(element) == PointsModel:
+            return element.compute().index.to_numpy()
```

Comment:
Same here but for `get_element_instances` in `relational_query`.

Reply:
done